contracts/src/CrossChainApplications/NativeTokenBridge/ERC20TokenSource.sol

```diff
@@ -0,0 +1,226 @@
+        require(
+            recipient != address(0),
+            "ERC20TokenSource: zero recipient address"
+        );
+
+        // Lock tokens in this contract. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedAmount = SafeERC20TransferFrom.safeTransferFrom(
+            IERC20(erc20ContractAddress),
+            totalAmount
+        );
+
+        // Ensure that the adjusted amount is greater than the fee to be paid.
+        require(
+            adjustedAmount > feeAmount,
+            "ERC20TokenSource: insufficient adjusted amount"
+        );
+
+        // Allow the Teleporter messenger to spend the fee amount.
+        if (feeAmount > 0) {
+            SafeERC20.safeIncreaseAllowance(
+                IERC20(erc20ContractAddress),
+                address(teleporterMessenger),
+                feeAmount
+            );
+        }
+
+        uint256 transferAmount = adjustedAmount - feeAmount;
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationChainID: destinationBlockchainID,
+                destinationAddress: nativeTokenDestinationAddress,
+                feeInfo: TeleporterFeeInfo({
+                    feeTokenAddress: erc20ContractAddress,
+                    amount: feeAmount
+                }),
+                requiredGasLimit: MINT_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(recipient, transferAmount)
+            })
+        );
+
+        emit TransferToDestination({
+            sender: msg.sender,
+            recipient: recipient,
+            transferAmount: transferAmount,
+            feeAmount: feeAmount,
+            teleporterMessageID: messageID
+        });
+    }
+
+    /**
+     * @dev Unlocks tokens to recipient.
+     */
+    function _unlockTokens(address recipient, uint256 amount) private {
+        require(
+            recipient != address(0),
+            "ERC20TokenSource: zero recipient address"
+        );
+
+        // Transfer to recipient
+        SafeERC20.safeTransfer(IERC20(erc20ContractAddress), recipient, amount);
+
+        emit UnlockTokens(recipient, amount);
+    }
+
+    /**
+     * @dev Sends tokens to BLACKHOLE_ADDRESS.
+     */
+    function _burnTokens(uint256 amount) private {
+        SafeERC20.safeTransfer(
+            IERC20(erc20ContractAddress),
+            BURNED_TX_FEES_ADDRESS,
+            amount
+        );
+        emit BurnTokens(amount);
+    }
+
+    /**
+     * @dev Update destinationChainBurnedBalance sent from destination chain
+     */
+    function _updateDestinationChainBurnedBalance(
+        uint256 newBurnBalance
+    ) private {
+        if (newBurnBalance > destinationChainBurnedBalance) {
```

Comment:
should `newBurnBalance` ever be less than `destiantionChainBurnedBalance`? Do we want to rvert if it is

Reply:
I don't think we want to revert here. See https://github.com/ava-labs/teleporter/pull/40#discussion_r1402571215